leader/leader.go

```diff
@@ -27,76 +27,93 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/wait"
 	crclient "sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/controller-runtime/pkg/client/config"
+	crconfig "sigs.k8s.io/controller-runtime/pkg/client/config"
 	logf "sigs.k8s.io/controller-runtime/pkg/log"
 )
 
-type runModeType string
-
-const (
-	localRunMode runModeType = "local"
-)
-
-// forceRunModeEnv indicates if the operator should be forced to run in either local
-// or cluster mode (currently only used for local mode)
-var forceRunModeEnv = "OSDK_FORCE_RUN_MODE"
-
-// errNoNamespace indicates that a namespace could not be found for the current
+// ErrNoNamespace indicates that a namespace could not be found for the current
 // environment
-var errNoNamespace = fmt.Errorf("namespace not found for current environment")
-
-// errRunLocal indicates that the operator is set to run in local mode (this error
-// is returned by functions that only work on operators running in cluster mode)
-var errRunLocal = fmt.Errorf("operator run mode forced to local")
+var ErrNoNamespace = fmt.Errorf("namespace not found for current environment")
 
 // podNameEnvVar is the constant for env variable POD_NAME
 // which is the name of the current pod.
 const podNameEnvVar = "POD_NAME"
 
+var readNamespace = func() ([]byte, error) {
+	return ioutil.ReadFile("/var/run/secrets/kubernetes.io/serviceaccount/namespace")
+}
+
 var log = logf.Log.WithName("leader")
 
 // maxBackoffInterval defines the maximum amount of time to wait between
 // attempts to become the leader.
 const maxBackoffInterval = time.Second * 16
 
+type Option func(*config) error
```

Comment:
Ah crap. I totally messed this up. Now that I'm seeing this, I feel like it's a little user-hostile. If we don't export `Config`, users can't provide their own `Option` implementations.

Given that we only have one field to set, it might not seem important now, but if we add more fields in the future, we want to give users the flexibility to write their own `Option`s. 

Opening this up later would be a breaking change to the `Option` definition.

So given all of this, I think we want:
```go
type Option func(*Config) error

type Config struct { ... }

// now that Config is exported, setDefaults would need 
// to be unexported since this only needs to be called
// by us after all options have run.
func (c *Config) setDefaults() error { ... }
```

Sorry for the back and forth!

Reply:
Reverted the commit that unexported Config. I then added a new commit to unexport just `setDefaults`